docs/api/session.md

```diff
@@ -1330,18 +1330,44 @@ the initial state will be `interrupted`. The download will start only when the
 
 Returns `Promise<void>` - resolves when the session’s HTTP authentication cache has been cleared.
 
-#### `ses.setPreloads(preloads)`
+#### `ses.setPreloads(preloads)` _Deprecated_
 
 * `preloads` string[] - An array of absolute path to preload scripts
 
 Adds scripts that will be executed on ALL web contents that are associated with
 this session just before normal `preload` scripts run.
 
-#### `ses.getPreloads()`
+**Deprecated:** Use the new `ses.registerPreloadScript` API. This will overwrite any preload scripts
+registered for `service-worker` context types.
+
+#### `ses.getPreloads()` _Deprecated_
 
 Returns `string[]` an array of paths to preload scripts that have been
 registered.
 
+**Deprecated:** Use the new `ses.getPreloadScripts` API. This will only return preload script paths
+for `frame` context types.
+
+#### `ses.registerPreloadScript(script)`
+
+* `script` [PreloadScriptRegistration](structures/preload-script-registration.md) - Preload script
+
+Registers preload script that will be executed in its associated context type in this session. For
+`frame` contexts, this will run prior to any preload defined in the web preferences of a
+WebContents.
+
+Returns `string` - The ID of the registered preload script.
+
+#### `ses.unregisterPreloadScript(id)`
```

Comment:
Where did we land on the alternative API here:

```
registerPreloadScript(path, opts);
unregisterPreloadSCript(path[, opts]);
```

I don't love this ID system (as discussed in API WG), this id serves no purpose other than a "pointer" which we don't need if we treat the path as a key to multiple scripts. `Record<Path, Opts[]>`

Reply:
Folks who reviewed the RFC (https://github.com/electron/rfcs/pull/8) seemed to be in support of this design. In my opinion, it feels idiomatic to common DOM APIs such as `setInterval` or `setTimeout`. Part of the inspiration was [registerContentScripts](https://developer.chrome.com/docs/extensions/reference/api/scripting#method-registerContentScripts) which also requires an `id` property. Recently I found that ControlledFrame's [addContentScript](https://wicg.github.io/controlled-frame/#dictdef-contentscriptdetails) also requires a `name`.

I took @itsananderson's suggestion of making the `id` optional ([5c8e11a](https://github.com/electron/electron/pull/44411/commits/5c8e11aa2ec30491b0a04fb48a743b693a214822)) which allows folks to only need to care about it if they want to remove it. I think this addresses the initial feedback brought up in the API WG meeting.

I'm hesitant to make `path` the key for a few reasons:
- Preload script don't technically require a file. In the future we could accept a `func` like with `executeScriptInMainWorld`. Or we extend it to allow preloading CSS as with Chrome extension's `registerContentScripts`, as an example. This forces us into a box that would require a full deprecation if we change this in the future.
- An `id` can be referenced by a constant which makes code maintenance simple. With a path, developers will need to worry about whether the app is packaged, what builder they're using, etc.
- `Record<Path, Opts[]>` feels awkward where a script would want to run in multiple contexts.